src/utils/compressor.cpp

```diff
@@ -0,0 +1,125 @@
+DEFINE_VALIDATED_string(storage_property_store_compression_level, "mid",
+                        memgraph::flags::storage_property_store_compression_level_help_string.c_str(),
+                        { return memgraph::flags::ValidStoragePropertyStoreCompressionLevel(value); });
+
+namespace memgraph::flags {
+
+bool ValidStoragePropertyStoreCompressionLevel(std::string_view value) {
+  if (const auto result = memgraph::utils::IsValidEnumValueString(value, compression_level_mappings);
+      result.HasError()) {
+    const auto error = result.GetError();
+    switch (error) {
+      case memgraph::utils::ValidationError::EmptyValue: {
+        std::cout << "Compression level cannot be empty." << '\n';
+        break;
+      }
+      case memgraph::utils::ValidationError::InvalidValue: {
+        std::cout << "Invalid value for compression level. Allowed values: "
+                  << memgraph::utils::GetAllowedEnumValuesString(compression_level_mappings) << '\n';
+        break;
+      }
+    }
+    return false;
+  }
+
+  return true;
+}
+
+int StoragePropertyStoreCompressionLevelToInt(std::string_view value) {
+  return memgraph::utils::StringToEnum<int>(value, compression_level_mappings).value();
+}
+
+}  // namespace memgraph::flags
+
+namespace memgraph::utils {
+
+DataBuffer::DataBuffer(uint8_t *data, uint32_t compressed_size, uint32_t original_size)
+    : data(data), compressed_size(compressed_size), original_size(original_size) {}
+
+DataBuffer::DataBuffer(std::unique_ptr<uint8_t[]> data, uint32_t compressed_size, uint32_t original_size)
+    : data(std::move(data)), compressed_size(compressed_size), original_size(original_size) {}
+
+DataBuffer::DataBuffer(DataBuffer &&other) noexcept
+    : data(std::move(other.data)),
+      compressed_size(std::exchange(other.compressed_size, 0)),
+      original_size(std::exchange(other.original_size, 0)) {}
+
+DataBuffer &DataBuffer::operator=(DataBuffer &&other) noexcept {
+  if (this != &other) {
+    data = std::move(other.data);
+    compressed_size = std::exchange(other.compressed_size, 0);
+    original_size = std::exchange(other.original_size, 0);
+  }
+  return *this;
+}
+
+DataBuffer ZlibCompressor::Compress(const uint8_t *input, uint32_t original_size) {
+  if (original_size == 0 || input == nullptr) {
+    return {};
+  }
+
+  auto compress_bound = compressBound(original_size);
+  std::unique_ptr<uint8_t[]> compressed_data(new uint8_t[compress_bound]);
+  const auto compressed_bound_copy = compress_bound;
+
+  const int result = compress2(
+      compressed_data.get(), &compress_bound, input, original_size,
+      memgraph::flags::StoragePropertyStoreCompressionLevelToInt(FLAGS_storage_property_store_compression_level));
+
+  if (result == Z_OK) {
+    std::unique_ptr<uint8_t[]> compressed_buffer_data;
+    const uint32_t compressed_size = compress_bound;
+    if (compress_bound < compressed_bound_copy) {
```

Comment:
Why do you need this check?

Reply:
it was because `compressed_bound_copy` was a guess of the compressed size, the value of `compress_bound` was changed by the call to `compress2`. If they are one and the same then we can use the allocated buffer as is. If not then we need to allocate a smaller buffer to shrink to only what was needed.

Code has been refactored so this is more obvious without need for a comment.